include/dicom/dicom.h

```diff
@@ -1479,91 +1635,167 @@ DcmFilehandle *dcm_filehandle_create_from_memory(DcmError **error,
  * :param filehandle: File
- * :param metadata: Metadata
  *
- * :return: Basic Offset Table
+ * :return: metadata
  */
 DCM_EXTERN
-DcmBOT *dcm_filehandle_read_bot(DcmError **error, DcmFilehandle *filehandle,
-                                DcmDataSet *metadata);
+const DcmDataSet *dcm_filehandle_get_metadata(DcmError **error,
+                                              DcmFilehandle *filehandle);
 
 /**
- * Build Basic Offset Table for a File.
+ * Read everything necessary to fetch frames from the file.
+ *
+ * Scans the PixelData sequence and loads the PerFrameFunctionalGroupSequence,
+ * if present.
+ *
+ * This function will be called automatically on the first call to
+ * :c:func:`dcm_filehandle_read_frame_position()` or
+ * :c:func:`dcm_filehandle_read_frame()`. It can take some time to execute,
+ * so it is available as a separate function call in case this delay needs
+ * to be managed.
+ *
+ * After calling this function, the filehandle read point is always
+ * positioned at the PixelData tag.
+ *
+ * It is safe to call this function many times.
  *
  * :param error: Pointer to error object
  * :param filehandle: File
- * :param metadata: Metadata
  *
- * :return: Basic Offset Table
+ * :return: true on success
  */
 DCM_EXTERN
-DcmBOT *dcm_filehandle_build_bot(DcmError **error, DcmFilehandle *filehandle,
-                                 DcmDataSet *metadata);
+bool dcm_filehandle_read_pixeldata(DcmError **error,
+                                   DcmFilehandle *filehandle);
 
 /**
  * Read an individual Frame from a File.
  *
+ * Frames are numbered from 1 in the order they appear in the PixelData element.
+ *
  * :param error: Pointer to error object
  * :param filehandle: File
- * :param metadata: Metadata
- * :param bot: Basic Offset Table
- * :param index: One-based offset of the Frame in the Pixel Data Element
+ * :param index: One-based frame number
  *
  * :return: Frame
  */
 DCM_EXTERN
 DcmFrame *dcm_filehandle_read_frame(DcmError **error,
                                     DcmFilehandle *filehandle,
-                                    DcmDataSet *metadata,
-                                    DcmBOT *bot,
-                                    uint32_t index);
+                                    uint32_t frame_number);
+
+/**
+ * Read the frame at a position in a File.
+ *
+ * Read a frame from a File at a specified (column, row), numbered from zero.
+ * This takes account of any frame positioning given in
+ * PerFrameFunctionalGroupSequence.
+ *
+ * :param error: Pointer to error object
+ * :param filehandle: File
+ * :param column: Column number, from 0
+ * :param row: Row number, from 0
+ *
+ * :return: Frame
+ */
+DCM_EXTERN
+DcmFrame *dcm_filehandle_read_frame_position(DcmError **error,
+                                             DcmFilehandle *filehandle,
+                                             uint32_t column,
+                                             uint32_t row);
```

Comment:
This assumes that the DICOM object represents a tiled image.

I suggest renaming to `dcm_filehandle_read_frame_at_tile_position` to make that more clear. I also suggest implementing a check to confirm that this assumption is met.

Reply:
It'll work for non-tiled images too -- just set col, row to (0, 0).

Reply:
Hmm...I find this behavior surprising, because non-tiled images don't really have any columns or rows. I would find it cleaner to implement a `dcm_filehandle_is_tiled_image` function to check whether the DICOM object is a tiled image (we may also want `dcm_filehandle_is_image`, which can be called internally) and fails if it isn't.

See corresponding Python implementation: https://github.com/ImagingDataCommons/dicomslide/blob/dc63981c5ac5a3a49e74e9ed373d06d5ebd3d46e/src/dicomslide/utils.py#L34

Reply:
I imagined a one-frame image (perhaps the label, for example) as a 1x1 array, so `dcm_filehandle_read_frame_position()` can make sense, I think. It makes openslide a bit simpler:

https://github.com/openslide/openslide/blob/b605126b8994b2748749e020d8126d1114667ce7/src/openslide-vendor-dicom.c#L578-L583

You can see the get-associated-image-pixel-data function can just call off into the main fetch-a-tile, it doesn't need a special read-untiled function, and it doesn't need an is-tiled tester.

Reply:
We should design the library not just with OpenSlide in mind. In my opinion, the library should be useful beyond whole slide imaging. Indexing image frames at a (column, row) position only makes sense for tiled images. If the image is not tiled, there is per definition no tile to be found at (0, 0) position.

OpenSlide will need to check for the image type and handle "associated images" differently.

Reply:
Openslide was just an example use case -- I think all programs would benefit from a single code path for both tiled and untiled images.

It checks that (column, row) is (0, 0) for untiled images, so I think it's safe and reasonable. Users can always call `read_frame(0)` if they wish.

Reply:
I agree John, let's close this discussion so that we can release this, unless anyone strenuously objects.